django/db/models/sql/compiler.py

```diff
@@ -1289,25 +1290,27 @@ def as_sql(self):
                 for p, vals in zip(placeholder_rows, param_rows)
             ]
 
-    def execute_sql(self, return_id=False):
-        assert not (
-            return_id and len(self.query.objs) != 1 and
-            not self.connection.features.can_return_ids_from_bulk_insert
+    def execute_sql(self):
+        return_id = (
+            len(self.query.objs) == 1 or
+            self.connection.features.can_return_rows_from_bulk_insert
         )
-        self.return_id = return_id
         with self.connection.cursor() as cursor:
             for sql, params in self.as_sql():
                 cursor.execute(sql, params)
-            if not return_id:
-                return
-            if self.connection.features.can_return_ids_from_bulk_insert and len(self.query.objs) > 1:
-                return self.connection.ops.fetch_returned_insert_ids(cursor)
-            if self.connection.features.can_return_id_from_insert:
+            if not (return_id and self.query.get_meta().returning_fields):
+                return []
+            if self.connection.features.can_return_rows_from_bulk_insert and len(self.query.objs) > 1:
+                return self.connection.ops.fetch_returned_insert_rows(cursor)
+            if self.connection.features.can_return_columns_from_insert:
                 assert len(self.query.objs) == 1
-                return self.connection.ops.fetch_returned_insert_id(cursor)
-            return self.connection.ops.last_insert_id(
-                cursor, self.query.get_meta().db_table, self.query.get_meta().pk.column
-            )
+                return self.connection.ops.fetch_returned_insert_columns(cursor)
+            if self.query.get_meta().pk.get_internal_type() in ['AutoField', 'BigAutoField']:
+                # The PK could be a custom field. Only AutoFields are supported in MySQL and SQLite.
```

Comment:
Something along the lines of:
```
For databases which do not support returning clause we need to ask the database for the id. Generally last_insert_id is only supported for serial/auto_incr columns, hence we guard it for auto field.
# TODO: Maybe add a marker to fields that their value can be returned by last_insert_id instead of a type check.
```